go.sum

```diff
@@ -591,8 +590,9 @@ github.com/spf13/cast v1.3.1 h1:nFm6S0SMdyzrzcmThSipiEubIDy8WEXKNZ0UOgiRpng=
 github.com/spf13/cast v1.3.1/go.mod h1:Qx5cxh0v+4UWYiBimWS+eyWzqEqokIECu5etghLkUJE=
 github.com/spf13/cobra v0.0.3/go.mod h1:1l0Ry5zgKvJasoi3XT1TypsSe7PqH0Sj9dhYf7v3XqQ=
 github.com/spf13/cobra v1.0.0/go.mod h1:/6GTrnGXV9HjY+aR4k0oJ5tcvakLuG6EuKReYlHNrgE=
-github.com/spf13/cobra v1.2.1 h1:+KmjbUw1hriSNMF55oPrkZcb27aECyrj8V2ytv7kWDw=
 github.com/spf13/cobra v1.2.1/go.mod h1:ExllRjgxM/piMAM+3tAZvg8fsklGAf3tPfi+i8t68Nk=
+github.com/spf13/cobra v1.5.0 h1:X+jTBEBqF0bHN+9cSMgmfuvv2VHJ9ezmFNf9Y/XstYU=
```

Comment:
This doesn't introduce https://github.com/docker/cli/issues/3621 ?

Reply:
Updated bench results in https://github.com/crazy-max/docker-cli-bench to take that PR into account as well as latest changes from docker/cli (docker/cli@f1615facb1ca44e4336ab20e621315fc2cfb845a) and it looks like it's actually slightly improved compared to `v0.8.2`:

`docker buildx version`:
* `v0.8.2` ~392ms
* `crazy-max/buildx#update-docker-cli` ~364ms

But a bit worse compared to current changes on master branch:

`docker buildx version`:
* `docker/buildx#master` ~352ms
* `crazy-max/buildx#update-docker-cli` ~364ms

Full bench:

```console
BuildxVersion/buildx-0-v0.8.2-docker-19.03.15-32                                                                            524ms ± 0%
BuildxVersion/buildx-0-v0.8.2-docker-20.10.0-32                                                                             482ms ± 0%
BuildxVersion/buildx-0-v0.8.2-docker-20.10.12-32                                                                            489ms ± 0%
BuildxVersion/buildx-0-v0.8.2-docker-20.10.17-32                                                                            488ms ± 0%
BuildxVersion/buildx-0-v0.8.2-docker-22.xx-0-2b4ffb301bb1cdd1adeb321807e10a9c656f346e-32                                    295ms ± 0%
BuildxVersion/buildx-0-v0.8.2-docker-22.xx-1-a09e61a247656c780d2ae487a276ea1761ceb85c-32                                    383ms ± 0%
BuildxVersion/buildx-0-v0.8.2-docker-22.xx-2-84b86e23b7f520b0bb956df88331d6c8d472ce0e-32                                    362ms ± 0%
BuildxVersion/buildx-0-v0.8.2-docker-22.xx-3-c59773f1551a8fd289538efc82274332f31f8c19-32                                    390ms ± 0%
BuildxVersion/buildx-0-v0.8.2-docker-22.xx-4-f1615facb1ca44e4336ab20e621315fc2cfb845a-32                                    392ms ± 0%
BuildxVersion/buildx-1-701c548e46348da2958104907c9572ea7ce6ab52-docker-19.03.15-32                                          491ms ± 0%
BuildxVersion/buildx-1-701c548e46348da2958104907c9572ea7ce6ab52-docker-20.10.0-32                                           456ms ± 0%
BuildxVersion/buildx-1-701c548e46348da2958104907c9572ea7ce6ab52-docker-20.10.12-32                                          474ms ± 0%
BuildxVersion/buildx-1-701c548e46348da2958104907c9572ea7ce6ab52-docker-20.10.17-32                                          457ms ± 0%
BuildxVersion/buildx-1-701c548e46348da2958104907c9572ea7ce6ab52-docker-22.xx-0-2b4ffb301bb1cdd1adeb321807e10a9c656f346e-32  275ms ± 0%
BuildxVersion/buildx-1-701c548e46348da2958104907c9572ea7ce6ab52-docker-22.xx-1-a09e61a247656c780d2ae487a276ea1761ceb85c-32  350ms ± 0%
BuildxVersion/buildx-1-701c548e46348da2958104907c9572ea7ce6ab52-docker-22.xx-2-84b86e23b7f520b0bb956df88331d6c8d472ce0e-32  351ms ± 0%
BuildxVersion/buildx-1-701c548e46348da2958104907c9572ea7ce6ab52-docker-22.xx-3-c59773f1551a8fd289538efc82274332f31f8c19-32  359ms ± 0%
BuildxVersion/buildx-1-701c548e46348da2958104907c9572ea7ce6ab52-docker-22.xx-4-f1615facb1ca44e4336ab20e621315fc2cfb845a-32  352ms ± 0%
BuildxVersion/buildx-2-update-docker-cli-docker-19.03.15-32                                                                 494ms ± 0%
BuildxVersion/buildx-2-update-docker-cli-docker-20.10.0-32                                                                  444ms ± 0%
BuildxVersion/buildx-2-update-docker-cli-docker-20.10.12-32                                                                 488ms ± 0%
BuildxVersion/buildx-2-update-docker-cli-docker-20.10.17-32                                                                 452ms ± 0%
BuildxVersion/buildx-2-update-docker-cli-docker-22.xx-0-2b4ffb301bb1cdd1adeb321807e10a9c656f346e-32                         272ms ± 0%
BuildxVersion/buildx-2-update-docker-cli-docker-22.xx-1-a09e61a247656c780d2ae487a276ea1761ceb85c-32                         354ms ± 0%
BuildxVersion/buildx-2-update-docker-cli-docker-22.xx-2-84b86e23b7f520b0bb956df88331d6c8d472ce0e-32                         361ms ± 0%
BuildxVersion/buildx-2-update-docker-cli-docker-22.xx-3-c59773f1551a8fd289538efc82274332f31f8c19-32                         374ms ± 0%
BuildxVersion/buildx-2-update-docker-cli-docker-22.xx-4-f1615facb1ca44e4336ab20e621315fc2cfb845a-32                         364ms ± 0%
```

Reply:
Any idea what changed between v0.8 and master?

What about the times in buildx itself. I assume in your times the main bottleneck is in compose-cli to docker-cli to buildx redirect and not in buildx itself.

Is the difference because buildx doesn't enable completions?

Reply:
> Any idea what changed between v0.8 and master?

I will bisect and keep you posted

> What about the times in buildx itself. I assume in your times the main bottleneck is in compose-cli to docker-cli to buildx redirect and not in buildx itself.

Here is the standalone bench result:

```
BuildxVersionStandalone/buildx-0-v0.8.2-32                                                                                  489ms ± 0%
BuildxVersionStandalone/buildx-1-701c548e46348da2958104907c9572ea7ce6ab52-32                                                456ms ± 0%
BuildxVersionStandalone/buildx-2-update-docker-cli-32                                                                       457ms ± 0%
```

Complete bench results: https://github.com/crazy-max/docker-cli-bench#readme

> Is the difference because buildx doesn't enable completions?

Good point will take a look at this

Reply:
Added `v0.7.0` to bench results: https://github.com/crazy-max/docker-cli-bench#readme

```
BuildxVersionStandalone/buildx-0-v0.7.0-32                                      502ms ± 0%
BuildxVersionStandalone/buildx-1-v0.8.2-32                                      524ms ± 0%
BuildxVersionStandalone/buildx-2-701c548e46348da2958104907c9572ea7ce6ab52-32    501ms ± 0%
BuildxVersionStandalone/buildx-3-update-docker-cli-32                           492ms ± 0%
```

I will run bench tests on a barebone system without Docker Desktop.